core/src/main/scala/hedgehog/package.scala

```diff
@@ -31,18 +31,86 @@ package object hedgehog extends ApplicativeSyntax {
   def propertyT: PropertyTOps =
     new PropertyTOps {}
 
+  /**
+    * Compare two arguments with the comparison function and return Result.success
+    * if the function return true. Otherwise, it returns Result.failure with
+    * the Log containing the argument values.
+    *
+    * @example
+    * {{{
+    *   val a1 = "abc"
+    *   val a2 = "abc"
+    *   diff(a1, a2)(_ == _)
+    *   // Result.success
+    *
+    *   diff(123, 456)(_ != _).log("It must be different.")
+    *   // Result.success
+    *
+    *   val x = 'q'
+    *   val y = 80
+    *   diff(x, y)((x, y) => y < 87 && x <= 'r')
+    *   // Result.success
+    * }}}
+    * @example
+    * {{{
+    *   val a1 = "abc"
+    *   val a2 = "xyz"
+    *   diff(a1, a2)(_ == _)
+    *   // Result.failure
+    *   > === Failed ===
+    *   > --- lhs ---
+    *   > abc
+    *   > --- rhs ---
+    *   > xyz
+    *
+    *   diff(123, 123)(_ != _).log("It must be different.")
+    *   // Result.failure
+    *   > === Failed ===
+    *   > --- lhs ---
+    *   > 123
+    *   > --- rhs ---
+    *   > 123
+    *   > It must be different.
+    *
+    *   val x = 'z'
+    *   val y = 100
+    *   diff(x, y)((x, y) => y < 87 && x <= 'r')
+    *   // Result.failure
+    *   > === Failed ===
+    *   > --- lhs ---
+    *   > z
+    *   > --- rhs ---
+    *   > 100
+    * }}}
+    *
+    * @param a1 The left-hand side argument
+    * @param a2 The right-hand side argument
+    * @param f The comparison function to compare the given two arguments
+    * @tparam A the type of the arguments
+    * @return Result.success. if the f(a1, a2) returns true.
+    *         Otherwise Result.failure with the Log containing the argument values
+    * @see https://github.com/hedgehogqa/haskell-hedgehog/blob/921e4af72a181f01d90816fd7055b823bf885b3b/hedgehog/src/Hedgehog/Internal/Property.hs#L707
+    */
+  def diff[A](a1: => A, a2: => A)(f: (A, A) => Boolean): Result =
+    diffNamed("=== Failed ===", a1, a2)(f)
+
+  def diffNamed[A](logName: String, a1: => A, a2: => A)(f: (A, A) => Boolean): Result = {
+    lazy val x = a1
+    lazy val y = a2
+    Result.assert(f(x, y))
+      .log(logName)
+      .log("--- lhs ---")
+      .log(String.valueOf(x))
+      .log("--- rhs ---")
+      .log(String.valueOf(y))
```

Comment:
I needed to use `String.valueOf()` as I get `NullPointerException` for the following code when x or y is `null`.

```scala
    Result.assert(f(x, y))
      .log(logName)
      .log("--- lhs ---")
      .log(x.toString)
      .log("--- rhs ---")
      .log(y.toString)
```
```scala
val a1: String = null
val a2: String = "abc"
diff(a1, a2)(_ == _)
```
```
[error] java.lang.NullPointerException
[error] 	at sbt.JUnitXmlTestsListener$TestSuite.$anonfun$stop$2(JUnitXmlTestsListener.scala:114)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
````